CONTRIBUTING.md

```diff
@@ -587,6 +590,61 @@ that are part of this project but not production code. The canonical example
+appropriate branches ourselves.
+
+We really appreciate everyone who is interested in contributing to
+Elasticsearch and regret that we sometimes have to reject contributions even
+when they might appear to make genuine improvements to the system. Reviewing
+contributions can be a very time-consuming task, yet the team is small and our
+time is very limited. In some cases the time we would need to spend on reviews
+would outweigh the benefits of a change by preventing us from working on other
+more beneficial tasks instead.
+
+Please discuss your change in a Github issue before spending much time on its
+implementation. We sometimes have to reject contributions that duplicate other
+efforts, take the wrong approach to solving a problem, or solve a problem which
+does not need solving. An up-front discussion would have saved a good deal of
+wasted time in these cases.
+
+We normally immediately reject isolated PRs that only perform simple
+refactorings or otherwise "tidy up" certain aspects of the code. We think the
+benefits of this kind of change are very small, and in our experience it is not
+worth investing the substantial effort needed to review them. This especially
+includes changes suggested by tools.
+
+We sometimes reject contributions due to the low quality of the submission
+since low-quality submissions tend to take unreasonable effort to review
+properly. Quality is rather subjective so it is hard to describe exactly how to
+avoid this, but there are some basic steps you can take to reduce the chances
+of rejection. Follow the guidelines listed above when preparing your changes.
+You should add tests that correspond with your changes, and your PR should pass
+affected test suites too. It makes it much easier to review if your code is
+formatted correctly and does not include unnecessary extra changes.
+
+We sometimes reject contributions if we find ourselves performing many review
+iterations without making enough progress. Some iteration is expected,
+particularly on technically complicated changes, and there's no fixed limit on
+the acceptable number of review cycles since it depends so much on the nature
+of the change. You can help to reduce the number of iterations by reviewing
+your contribution yourself or in your own team before asking us for a review.
+You may be surprised how many comments you can anticipate and address by taking
+a short break and then carefully looking over your changes again.
+
+We expect you to follow up on review comments somewhat promptly, but recognise
+that everyone has many priorities for their time and may not be able to respond
+for several days. We will understand if you find yourself without the time to
+complete your contribution, but please let us know that you have stopped
+working on it. We will try to send you a reminder if we haven't heard from you
+in a while, but may end up closing your PR if you do not respond for too long.
+
+
```

Comment:
I wonder if we should end with a note about appealing against rejected PRs? Something like:

```suggestion
If your contribution is rejected we will close the pull request with a comment explaining why it was rejected. If you feel we have misunderstood your intended change or otherwise feel that the contribution should be reconsidered please add a comment explaining why and we will happily reconsider.
```

Reply:
Nice idea, yes. I've reworded this a bit in bd6ad15, WDYT?